test/e2e/logcheck/logcheck.go

```diff
@@ -0,0 +1,391 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package logcheck extends a Ginkgo test suite with optional monitoring
+// of system log files. Right now it can detect DATA RACE problems.
+//
+// This may get expanded in the future, for example to check also
+// for unhandled and potentially unexpected errors (https://github.com/kubernetes/kubernetes/issues/122005)
+// or log spam (https://github.com/kubernetes/kubernetes/issues/109297).
+//
+// Because it grabs log output as it is produced, it is possible to dump *all*
+// data into $ARTIFACTS/system-logs, in contrast to other approaches which
+// potentially only capture the tail of the output (log rotation).
+//
+// The additional check(s) must be opt-in because they are not portable.
+// The package registers additional command line flags for that.
+// Dedicated jobs are necessary which enable them.
+//
+// For DATA RACE, the binaries in the cluster must be compiled with race
+// detection (https://github.com/kubernetes/kubernetes/pull/133834/files).
+// It's also necessary to specify which containers to check. Other output
+// (e.g. kubelet in a kind cluster) currently cannot be checked.
+package logcheck
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"maps"
+	"os"
+	"path"
+	"regexp"
+	"slices"
+	"strings"
+	"sync"
+
+	"github.com/onsi/ginkgo/v2"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	"k8s.io/klog/v2/textlogger"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/podlogs"
+)
+
+const (
+	// defaultNamespacesRE is the default regular expression for which namespaces need to be checked,
+	// if checking is enabled: anything which contains the word "system".
+	//
+	// If we had an API for identifying "system" namespaces, we could use that, but we
```

Comment:
This also won't get any rotated log files ... I still don't think the tests should be handling logs, the built in logs APIs only give us best effort for recent output, but we can do better than that on the clusters were create for testing, it's just specific to the tools used.

Reply:
One more thought on "grab logs after a test run" (no matter how): I tried that initially and found that I had to reduce logging verbosity to avoid getting output discarded. We could try to change the log rotation settings, but that opens another can of worms because we found that they [are load-bearing](https://github.com/kubernetes-sigs/kind/blob/c8778675066f2437874270dc998d1d73d487baa0/hack/ci/e2e-k8s.sh#L184-L189) .

I'd prefer to keep them as similar to other jobs as possible (same verbosity level, same log rotation).